feat: strongly-typed `Container`

alecgibson opened the pull request:
## Description

This is a non-breaking change, affecting only TypeScript types, and doesn't change the implementation in any way.

### Motivation

`inversify` already has some basic support for types when binding, and retrieving bindings.

However, the type support requires manual intervention from developers, and can be error-prone.

For example, the following code will happily compile, even though the types here are inconsistent:

```ts
container.bind<Bar>('bar').to(Bar);
const foo = container.get<Foo>('bar')
```

Furthermore, this proposal paves the way for [type-safe injection][1], which will be added once this change is in.

### Improved type safety

This change adds an optional type parameter to the `Container`, which takes an identifier map as an argument. For example:

```ts
type IdentifierMap = {
  foo: Foo;
  bar: Bar;
};

const container = new Container<IdentifierMap>;
```

If a `Container` is typed like this, we now get strong typing both when binding, and getting bindings:

```ts
const container = new Container<IdentifierMap>;

container.bind('foo').to(Foo); // ok
container.bind('foo').to(Bar); // error

const foo: Foo = container.get('foo') // ok
const bar: Bar = container.get('foo') // error
```

This also has the added benefit of no longer needing to pass around service identifier constants: the strings (or symbols) are all strongly typed, and will fail compilation if an incorrect one is used.

### Non-breaking

This change aims to make no breaks to the existing types, so any `Container` without an argument should continue to work as it did before.

[1]: https://github.com/inversify/InversifyJS/issues/788#issuecomment-2209341498

## Related Issue

https://github.com/inversify/InversifyJS/issues/788

## How Has This Been Tested?

- added tests to this test suite, which passes
- based on type definitions we've been using in Production ourselves

## Types of changes

- [ ] Updated docs / Refactor code / Added a tests case (non-breaking change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have updated the changelog.

alecgibson committed:
feat: strongly-typed `Container`

This is a non-breaking change, affecting only TypeScript types, and
doesn't change the implementation in any way.

Motivation
==========

`inversify` already has some basic support for types when binding, and
retrieving bindings.

However, the type support requires manual intervention from developers,
and can be error-prone.

For example, the following code will happily compile, even though the
types here are inconsistent:

```ts
container.bind<Bar>('bar').to(Bar);
const foo = container.get<Foo>('bar')
```

Furthermore, this paves the way for [type-safe injection][1], which will
be added once this change is in.

Improved type safety
====================

This change adds an optional type parameter to the `Container`, which
takes an identifier map as an argument. For example:

```ts
type IdentifierMap = {
  foo: Foo;
  bar: Bar;
};

const container = new Container<IdentifierMap>;
```

If a `Container` is typed like this, we now get strong typing both when
binding, and getting bindings:

```ts
const container = new Container<IdentifierMap>;

container.bind('foo').to(Foo); // ok
container.bind('foo').to(Bar); // error

const foo: Foo = container.get('foo') // ok
const bar: Bar = container.get('foo') // error
```

This also has the added benefit of no longer needing to pass around
service identifier constants: the strings (or symbols) are all strongly
typed, and will fail compilation if an incorrect one is used.

Non-breaking
============

This change aims to make no breaks to the existing types, so any
`Container` without an argument should continue to work as it did
before.

[1]: https://github.com/inversify/InversifyJS/issues/788#issuecomment-2209341498

alecgibson commented on src/interfaces/interfaces.ts:
```diff
@@ -226,18 +275,23 @@ namespace interfaces {
 
-  export type Bind = <T = unknown>(serviceIdentifier: ServiceIdentifier<T>) => BindingToSyntax<T>;
+  export type Bind<T extends BindingMap = any> =
+    <B extends ContainerBinding<T, K>, K extends ContainerIdentifier<T> = any>(serviceIdentifier: K) => BindingToSyntax<B>;
```
Note, in the interest of keeping changes to a minimum, I haven't touched the `BindingToSyntax` type, but we may want to consider tweaking it — either as part of this change, or a future change — because:

```ts
container.bind('foo').toDynamicValue(
  ({container}) => container.get('bar') // not strongly typed
)
```

notaphplover commented on src/interfaces/interfaces.ts:
```diff
@@ -46,7 +46,16 @@ namespace interfaces {
-  export type ServiceIdentifier<T = unknown> = (string | symbol | Newable<T> | Abstract<T>);
+  export type IfAny<T, Y, N> = 0 extends (1 & T) ? Y : N;
+
+  export type PropertyServiceIdentifier = (string | symbol);
```
This name is sort of misleading. A property can be decorated with any `ServiceIdentifier`. Is this a `BindingMapProperty`

notaphplover commented on src/interfaces/interfaces.ts:
```diff
@@ -46,7 +46,16 @@ namespace interfaces {
+  export type IfAny<T, Y, N> = 0 extends (1 & T) ? Y : N;
+
+  export type PropertyServiceIdentifier = (string | symbol);
+  export type ServiceIdentifier<T = unknown> = (PropertyServiceIdentifier | Newable<T> | Abstract<T>);
```
```suggestion
  export type ServiceIdentifier<T = unknown> = (string | symbol | Newable<T> | Abstract<T>);
```

notaphplover commented on src/interfaces/interfaces.ts:
```diff
@@ -46,7 +46,16 @@ namespace interfaces {
+  export type PropertyServiceIdentifier = (string | symbol);
+  export type ServiceIdentifier<T = unknown> = (PropertyServiceIdentifier | Newable<T> | Abstract<T>);
+  export type BindingMap = Record<PropertyServiceIdentifier, any>;
+  export type BindingMapKey<T extends BindingMap> = keyof T & PropertyServiceIdentifier;
```
This feels redundant. `keyof T & PropertyServiceIdentifier` should be equivalent to `keyof T` since T is enforced to have an index of `PropertyServiceIdentifier`. Am I missing something?

If so, lets remove this type in favor of `keyof` usages

alecgibson replied:
I agree it feels redundant, but if you try to change:

```diff
-export type ContainerIdentifier<T extends BindingMap> = IfAny<T, ServiceIdentifier, BindingMapKey<T>>;
+export type ContainerIdentifier<T extends BindingMap> = IfAny<T, ServiceIdentifier, keyof T>;
```

Then you get a slew of compilation errors, along the lines of:

```
src/container/container.ts:181:33 - error TS2345: Argument of type 'string | number | symbol | Newable<unknown> | Abstract<unknown>' is not assignable to parameter of type 'ServiceIdentifier<unknown>'.
  Type 'number' is not assignable to type 'ServiceIdentifier<unknown>'.

181     this._bindingDictionary.add(serviceIdentifier, binding as Binding<unknown>);
```

You can see from the error that `number` has snuck in as a possible identifier, which `ServiceIdentifier` doesn't allow.

I think this is from the possibility that the `BindingMap` could be set to `any` (if no map is provided, which is the backwards-compatible case). `any` can technically have `number` as a key, hence the error.

We need to keep the `any` possibility to allow for containers that use non-`string` or `symbol` identifiers, (eg class identifiers) which is incompatible with this `BindingMap` approach, since classes can't be property keys.

If you can find a way to get this to compile without this, then I'm all ears, but this definitely feels like the simplest solution to me.

notaphplover commented on src/interfaces/interfaces.ts:
```diff
@@ -46,7 +46,16 @@ namespace interfaces {
   }
 
-  export type ServiceIdentifier<T = unknown> = (string | symbol | Newable<T> | Abstract<T>);
+  export type IfAny<T, Y, N> = 0 extends (1 & T) ? Y : N;
```
Really smart, but sort of tricky. Tbh I don't see a better way to represent this.

notaphplover commented on src/interfaces/interfaces.ts:
```diff
@@ -46,7 +46,16 @@ namespace interfaces {
+  export type ServiceIdentifier<T = unknown> = (PropertyServiceIdentifier | Newable<T> | Abstract<T>);
+  export type BindingMap = Record<PropertyServiceIdentifier, any>;
+  export type BindingMapKey<T extends BindingMap> = keyof T & PropertyServiceIdentifier;
+  export type ContainerIdentifier<T extends BindingMap> = IfAny<T, ServiceIdentifier, BindingMapKey<T>>;
```
```suggestion
  export type MappedServiceIdentifier<T extends BindingMap> = IfAny<T, ServiceIdentifier, BindingMapKey<T>>;
```
It feels more a mapped service identifier, isn't it? I think we don't need to involve the concept of container.

notaphplover commented on src/interfaces/interfaces.ts:
```diff
@@ -46,7 +46,16 @@ namespace interfaces {
+  export type BindingMap = Record<PropertyServiceIdentifier, any>;
+  export type BindingMapKey<T extends BindingMap> = keyof T & PropertyServiceIdentifier;
+  export type ContainerIdentifier<T extends BindingMap> = IfAny<T, ServiceIdentifier, BindingMapKey<T>>;
+  export type ContainerBinding<T extends BindingMap, K extends ContainerIdentifier<T> = any> = K extends keyof T ? T[K] :
```
I know I'm super picky with names, sry about it. I understand this is the service type of a binding, but not a binding. Am I right?

alecgibson replied:
No worries. Naming stuff is the hardest part of the job! It's especially tricky because I'm not super well-versed in this library, so have probably just been naming stuff wrong.

This type `ContainerBinding<T, K>` is meant to represent the return value type of `container<T>.get(key: K): ContainerBinding<T, K>`. So I guess it's whatever you want to call that?

notaphplover replied:
Then I guess it's a `ServiceType` (`BindingMapServiceType` would be too long, wouldn't it be?)

notaphplover commented on test/interfaces/interfaces.test.ts:
```diff
@@ -0,0 +1,124 @@
+import { interfaces } from '../../src/interfaces/interfaces';
```
:heart: 

Thanks for adding this test module

notaphplover reviewed (commented):
This PR is sort of challenging to review. I would love to discuss some concepts before continue reviewing.

Overal, the PR looks amazing, I think it's going to provide so much value to the container!

PodaruDragos commented on src/interfaces/interfaces.ts:
```diff
@@ -46,7 +46,16 @@ namespace interfaces {
+
+  export type PropertyServiceIdentifier = (string | symbol);
+  export type ServiceIdentifier<T = unknown> = (PropertyServiceIdentifier | Newable<T> | Abstract<T>);
+  export type BindingMap = Record<PropertyServiceIdentifier, any>;
```
is there any way we can have this as `Record<PropertyServiceIdentifier, unknown>`; ? avoiding the `any` there i mean

alecgibson replied:
What's the benefit? I think `any` is more semantic here?

PodaruDragos replied:
no real benefit, keep consistency with most of the types since I tried really hard at some point to move away from most of `any`'s and have "proper" types

alecgibson replied:
I had a quick go, and it didn't immediately compile. I'm not personally sure it's worth investigating, because as I say the semantics aren't really "correct" (according to my limited understanding of `unknown` vs `any` 😅 ), since the values assigned in this `BindingMap` will be entirely consumer-generated, and they can legitimately set it to anything they want (ie it is semantically `any`, right...?).

If you feel strongly I can try to investigate, but the types don't really like it as-is:

```
src/container/container.ts:287:41 - error TS2345: Argument of type 'this' is not assignable to parameter of type 'Container<BindingMap, any>'.
  Type 'Container<T, P>' is not assignable to type 'Container<BindingMap, any>'.
    The types returned by 'bind(...)' are incompatible between these types.
      Type 'BindingToSyntax<ContainerBinding<T, K>>' is not assignable to type 'BindingToSyntax<B>'.
        Type 'ContainerBinding<T, K>' is not assignable to type 'B'.
          'ContainerBinding<T, K>' is assignable to the constraint of type 'B', but 'B' could be instantiated with a different subtype of constraint 'unknown'.
            Type 'ContainerBinding<T, string> | ContainerBinding<T, symbol>' is not assignable to type 'B'.
              'ContainerBinding<T, string> | ContainerBinding<T, symbol>' is assignable to the constraint of type 'B', but 'B' could be instantiated with a different subtype of constraint 'unknown'.
                Type 'ContainerBinding<T, string>' is not assignable to type 'B'.
                  'ContainerBinding<T, string>' is assignable to the constraint of type 'B', but 'B' could be instantiated with a different subtype of constraint 'unknown'.
                    Type 'T[keyof T & string]' is not assignable to type 'B'.
                      'T[keyof T & string]' is assignable to the constraint of type 'B', but 'B' could be instantiated with a different subtype of constraint 'unknown'.
                        Type 'T[string]' is not assignable to type 'B'.
                          'T[string]' is assignable to the constraint of type 'B', but 'B' could be instantiated with a different subtype of constraint 'unknown'.

287       const request = createMockRequest(this, serviceIdentifier, key, value);
```

PodaruDragos replied:
I don't really think you can go around this tbh. If you can that would be amazing. depends what @notaphplover thinks also

alecgibson committed:
review: rename new binding map types

 - `PropertyServiceIdentifier` -> `BindingMapProperty`
 - `ContainerIdentifier` -> `MappedServiceIdentifier`
 - revert `ServiceIdentifier` type to previous definition
